tools/sst_dump_tool.cc

```diff
@@ -182,6 +186,100 @@ Status SstFileDumper::DumpTable(const std::string& out_filename) {
   return s;
 }
 
+struct DumpLogReporter : public log::Reader::Reporter {
+  Status* status;
+  virtual void Corruption(size_t /*bytes*/, const Status& s) override {
+    if (this->status->ok()) *this->status = s;
+  }
+};
+
+Status SstFileDumper::RebuildTable(const std::string& out_filename,
+                                   const std::string& manifest_filename,
+                                   const std::string& out_manifest_filename) {
+  FileSystem* fs = FileSystem::Default().get();
+  {
+    std::unique_ptr<FSWritableFile> out_file;
+
+    Status s = NewWritableFile(fs, out_filename, &out_file, FileOptions());
+    if (!s.ok()) {
+      return s;
+    }
+
+    WritableFileWriter writer(std::move(out_file), out_filename, FileOptions());
+    s = table_reader_->RebuildTable(&writer, options_);
+    if (!s.ok()) {
+      return s;
+    }
+    s = writer.Close();
+    if (!s.ok()) {
+      return s;
+    }
+  }
+
+  uint64_t sstNumber;
+  FileType type;
+  WalFileType log_type;
+  if (!ParseFileName(file_name_, &sstNumber, &type, &log_type)) {
+    return Status::InvalidArgument(std::string("Cannot parse file name: ") +
+                                   file_name_);
+  }
+
+  std::unique_ptr<FSSequentialFile> file;
+  Status s =
+      fs->NewSequentialFile(manifest_filename, FileOptions(), &file, nullptr);
+  if (!s.ok()) {
+    return s;
+  }
+
+  std::unique_ptr<SequentialFileReader> file_reader(
+      new SequentialFileReader(std::move(file), manifest_filename, 1024));
+
+  std::unique_ptr<FSWritableFile> out_file_log;
+  s = NewWritableFile(fs, out_manifest_filename, &out_file_log, FileOptions());
+  std::unique_ptr<WritableFileWriter> writer_log(new WritableFileWriter(
+      std::move(out_file_log), out_manifest_filename, FileOptions()));
+
+  log::Writer logWriter(std::move(writer_log), 0, false);
+
+  DumpLogReporter reporter;
+  reporter.status = &s;
+  log::Reader reader(nullptr, std::move(file_reader), &reporter,
+                     true /* checksum */, 0 /* log_number */);
+  Slice record;
+  std::string scratch;
+  while (reader.ReadRecord(&record, &scratch) && s.ok()) {
+    VersionEdit edit;
+    s = edit.DecodeFrom(record);
+    if (!s.ok()) {
+      break;
+    }
+
+    for (std::vector<std::pair<int, FileMetaData>>::const_iterator it =
+             edit.GetNewFiles().cbegin();
+         it != edit.GetNewFiles().cend(); it++) {
+      if (it->second.fd.GetNumber() == sstNumber) {
```

Comment:
Can we append a manifest entry that does delete file followed by add file? Note other parts of the file entry may have changed like file endpoints in case the first or last block got dropped.

Reply:
Sure, sounds good, I will take a look.